File: expressjs/src/storage/storage.ts
import type { Readable } from "node:stream";

/** FileStorage abstracts file persistence. Local-disk today, S3 later. */
export interface FileStorage {
  /** Persist file content, return storage path for retrieval/deletion. */
  save(appName: string, fileID: string, filename: string, buffer: Buffer): Promise<string>;
  /** Open a readable stream for the stored file. */
  openStream(storagePath: string): Readable;
  /** Remove the file from storage. */
  delete(storagePath: string): Promise<void>;
}

File: backend/golang/internal/storage/storage.go
package storage

import (
	"context"
	"io"
)

// FileStorage abstracts file persistence. Local-disk today, S3 later.
type FileStorage interface {
	// Save persists file content and returns the storage path (used for retrieval/deletion).
	Save(ctx context.Context, appName, fileID, filename string, reader io.Reader) (storagePath string, err error)
	// Open returns a reader for the stored file.
	Open(ctx context.Context, storagePath string) (io.ReadCloser, error)
	// Delete removes the file from storage.
	Delete(ctx context.Context, storagePath string) error
}
